### PR TITLE
arm/cortex-a,r: replace cp15 instruct to macros to align operation

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arm/cortex-a,r: replace cp15 instruct to macros to align operation

This is continue work of https://github.com/apache/nuttx/pull/13486
Discussion here:
https://github.com/apache/nuttx/pull/13486#discussion_r1764354675

1. move cp15.h to arch public
2. replace cp15 instruct to macros to align operation
3. add memory barrier to avoid compiler optimization

Signed-off-by: chao an <anchao@lixiang.com>

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

**Comment by lupyuen** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13494&type=pullrequests)

Let's break down this PR against the NuttX requirements:

**Summary:**

* **Why is change necessary?** 
    * This part is a little weak. While it links to a prior PR and discussion, it should summarize the *reason* for those changes within this PR's description.  Is it a bug fix? Performance improvement? Code clarity? 
* **What functional part of the code is being changed?**
    *  Clearly identifies: `arm/cortex-a,r` architecture code, specifically related to `cp15` instructions.
* **How does the change work?**
    * Provides a good high-level overview:
        * Moves `cp15.h`
        * Replaces instructions with macros for alignment.
        * Adds memory barriers. 
* **Related Issues/PRs:** 
    * Links provided, which is excellent.

**Impact:**

* **Mostly "N/A" – This needs more detail.** Even if the impact is minimal, explain *why*.  For example:
    * **User Impact:** "No change in user-facing behavior".
    * **Build Impact:** "No changes to the build process itself, though compiled output may differ slightly".
    * **Hardware Impact:**  Specify if this is only relevant to *some* ARM/Cortex variants or if it's broad.
    * **Documentation:**  Does `cp15.h` being moved require any documentation updates?
    * **Security:**  Consider if using macros instead of raw instructions has any security implications (even if unlikely).
    * **Compatibility:**  Crucial for architecture-specific changes. Are there any compatibility concerns with older ARM versions?

**Testing:**

* **Build Hosts:**  Insufficient. List the *specific* OS, compiler versions, etc., used for testing.
* **Targets:** Be specific about the ARM/Cortex architectures and boards tested on, including simulator usage if any.
* **Testing logs:**  While "ci-check" implies CI was used, include relevant snippets of logs showing:
    * **Before:**  The issue/behavior the change addresses.
    * **After:**  The successful resolution or improvement.

**Overall:**

The PR has a good structure and clearly explains the technical *what*. However, it's weak on the *why* (justification) and needs more thorough impact assessment and testing details. 

**Recommendation:**  Don't just state "N/A" – explain *why* there's no impact for each category.  Provide specific testing information to build confidence in the change.

File: arch/arm/include/armv7-a/irq.h [hunk 1]
```diff
@@ -36,6 +36,9 @@
 #  include <stdint.h>
 #endif
 
+#include "cp15.h"
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:39]:
```suggestion
#include <arch/cp15.h>
```

File: arch/arm/include/armv7-a/irq.h [hunk 2]
```diff
@@ -36,6 +36,9 @@
...
+#include "barriers.h"
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:40]: ditto

File: arch/arm/include/armv7-a/irq.h [hunk 3]
```diff
@@ -457,11 +460,8 @@ static inline_function int up_cpu_index(void)
 
   /* Read the Multiprocessor Affinity Register (MPIDR) */
 
-  __asm__ __volatile__
-  (
-    "mrc " "p15, " "0" ", %0, " "c0" ", " "c0" ", " "5" "\n"
-    : "=r"(mpidr)
-  );
+  ARM_ISB();
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:463]: why add barrier for readonly register

File: arch/arm/include/armv7-a/irq.h [hunk 4]
```diff
@@ -500,23 +500,15 @@ static inline_function uint32_t up_getsp(void)
...
 static inline_function uint32_t *up_current_regs(void)
 {
-  uint32_t *regs;
-  __asm__ __volatile__
-  (
-    "mrc " "p15, " "0" ", %0, " "c13" ", " "c0" ", " "4" "\n"
-    : "=r"(regs)
-  );
-  return regs;
+  ARM_ISB();
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:503]: could you point out which document require to add barrier for TPIDRPRW read?

File: arch/arm/include/armv7-a/irq.h [hunk 5]
```diff
@@ -500,23 +500,15 @@ static inline_function uint32_t up_getsp(void)
...
 noinstrument_function
 static inline_function void up_set_current_regs(uint32_t *regs)
 {
-  __asm__ __volatile__
-  (
-    "mcr " "p15, " "0" ", %0, " "c13" ", " "c0" ", " "4" "\n"
-    :: "r"(regs)
-  );
+  CP15_SET(TPIDRPRW, regs);
+  ARM_ISB();
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-a/irq.h:511]: ditto

File: arch/arm/src/Makefile [hunk 6]
```diff
@@ -37,11 +37,12 @@ else                                    # ARM9, ARM7TDMI, etc.
...
-ARCH_SRCDIR = $(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)$(DELIM)src
+ARCH_SRCDIR = $(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)
 
-INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)chip
-INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)common
-INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)$(ARCH_SUBDIR)
+INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)src$(DELIM)chip
+INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)src$(DELIM)common
+INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)src$(DELIM)$(ARCH_SUBDIR)
+INCLUDES += ${INCDIR_PREFIX}$(ARCH_SRCDIR)$(DELIM)include$(DELIM)$(ARCH_SUBDIR)
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:45]: why need change

File: arch/arm/src/Makefile [hunk 7]
```diff
@@ -37,11 +37,12 @@ else                                    # ARM9, ARM7TDMI, etc.
 ARCH_SUBDIR = arm
 endif
 
-ARCH_SRCDIR = $(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)$(DELIM)src
+ARCH_SRCDIR = $(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:40]: how about cmake

File: arch/arm/include/armv7-a/irq.h (hunk 4 above)
Comment (anchao) [reply arch/arm/include/armv7-a/irq.h]: I just want to add the volatile attribute, it looks like it's already included in the CP15 operation, let me revert

File: arch/arm/src/Makefile (hunk 7 above)
Comment (anchao) [reply arch/arm/src/Makefile:40]: cmake has also changed

File: arch/arm/include/armv7-a/irq.h (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/arm/include/armv7-a/irq.h:39]: change to:
```
#include <arch/armv7-a/cp15.h>
```
the standard header file should be included through `<>`

File: arch/arm/src/Makefile (hunk 6 above)
Comment (xiaoxiang781216) [reply arch/arm/src/Makefile:45]: but we should use `<>` not `""` to include the header files which locates in the public folder.

File: arch/arm/include/armv7-a/irq.h (hunk 1 above)
Comment (anchao) [reply arch/arm/include/armv7-a/irq.h]: Done

File: arch/arm/src/armv8-r/arm_vectors.S [hunk 8]
```diff
@@ -23,10 +23,10 @@
  ****************************************************************************/
 
 #include <nuttx/config.h>
-#include <nuttx/irq.h>
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/arm_vectors.S:26]: don't need change nuttx/irq.h include arch/irq.h

File: arch/arm/src/armv8-r/arm_arch_timer.c [hunk 9]
```diff
@@ -34,7 +34,6 @@
 #include <nuttx/timers/arch_alarm.h>
 
 #include "barriers.h"
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/arm_arch_timer.c:36]: rermove

File: arch/arm/src/armv8-r/sctlr.h [hunk 10]
```diff
@@ -32,8 +32,9 @@
  * Included Files
  ****************************************************************************/
 
+#include <arch/irq.h>
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/sctlr.h:35]: don't need

Comment (anchao) [reply arch/arm/src/armv8-r/sctlr.h:35]: sctlr.h requires the relevant definition of cp15

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**